ExpressMapper NET40/MappingServiceBase.cs

```diff
@@ -423,28 +412,56 @@ protected LoopExpression CreateLoopExpression(Expression doMoveNextSrc, BlockExp
             {
                 return Expression.Call(destColl, destList.GetMethod("ToArray"));
             }
-            if (!destPropType.IsGenericType) return destColl;
+
+            if (!destPropType.IsGenericType && GetEnumerableInterface(destPropType) == null)
+            {
+                return destColl;
+            }
 
             if (typeof(IQueryable).IsAssignableFrom(destPropType))
             {
                 return Expression.Call(typeof(Queryable), "AsQueryable", new[] { destType }, destColl);
             }
 
-            if (destPropType.IsInterface && destColl.Type.IsSubclassOf(destPropType))
-            {
+            if (destPropType.IsInterface && destPropType.IsAssignableFrom(destColl.Type)) {
+                // This will handle any destination interface implemented by List<T>
                 return destColl;
             }
 
-            if (destPropType.IsClass)
-            {
+            ConstructorInfo ctor = null;
 
+            if (destPropType.IsInterface)
+            {
+                // We are targeting an interface type, we need to find a compatible collection type
+                // We could look for a loaded type that implements the target interface and has an appropriate
+                // constructor, but that is a bit too much magic for now.
+                throw new NotImplementedException(
+                    "Destination interface type " + destPropType.FullName + " is not supported yet" );

```

Comment:
First off, my PR was *not* intended to deal with non-generic IEnumerable etc, it was intended to deal with the case that a class implemented a *generic* collection interface but was not by itself generic (ie MySpecialList : IEnumerable<T>).

However, the code would already work for non-generic IEnumerable/IList etc as those are implemented by List<T> ( see https://msdn.microsoft.com/en-us/library/6sh2ey19(v=vs.110).aspx ) and therefore we would not reach this code as the check in line 426 would just return the list we already built.

My comment here was intended to cover the following scenario:

Assume we have destination type ICustomInterface<T> implementing IEnumerable<T>.
Now, we have to *search* the loaded assemblies for a type that implements ICustomInterface<T>. *That* is the magic i didnt implement :)